prql-compiler/src/codegen.rs

```diff
@@ -0,0 +1,590 @@
+            TransformCall(transform) => {
+                Some(format!("{} <unimplemented>", (*transform.kind).as_ref()))
+            }
+            Literal(literal) => Some(literal.to_string()),
+            Case(cases) => {
+                let mut r = String::new();
+                r += "case ";
+                r += &SeparatedExprs {
+                    exprs: cases,
+                    inline: ", ",
+                    line_end: ",",
+                }
+                .write_between("{", "}", opt)?;
+                Some(r)
+            }
+            BuiltInFunction { .. } => Some("<built-in>".to_string()),
+            Type(ty) => ty.write(opt),
+            Param(id) => Some(format!("${id}")),
+        }
+    }
+}
+
+/// Writes an optionally parenthesized expression
+fn write_expr(expr: &pl::Expr, parent: &pl::ExprKind, opt: WriteOpt) -> Option<String> {
+    let strength_self = binding_strength(&expr.kind);
+    let strength_parent = binding_strength(parent);
+
+    if strength_parent <= strength_self {
+        expr.write_between("(", ")", opt)
+    } else {
+        expr.write(opt)
+    }
+}
+
+fn binding_strength(expr: &pl::ExprKind) -> i32 {
+    match expr {
+        pl::ExprKind::Ident(_) => 1,
+        pl::ExprKind::All { .. } => 1,
+
+        pl::ExprKind::Range(_) => 3,
+        pl::ExprKind::Binary { op, .. } => match op {
+            pl::BinOp::Mul | pl::BinOp::Div | pl::BinOp::Mod => 4,
+            pl::BinOp::Add | pl::BinOp::Sub => 5,
+            pl::BinOp::Eq
+            | pl::BinOp::Ne
+            | pl::BinOp::Gt
+            | pl::BinOp::Lt
+            | pl::BinOp::Gte
+            | pl::BinOp::Lte
+            | pl::BinOp::RegexSearch => 6,
+            pl::BinOp::And => 8,
+            pl::BinOp::Or => 9,
+            pl::BinOp::Coalesce => 7,
+        },
+        pl::ExprKind::Unary { .. } => 2,
+        pl::ExprKind::FuncCall(_) => 10,
+        pl::ExprKind::Closure(_) => 10,
+
+        _ => 0,
+    }
+}
+
+impl WriteSource for pl::Ident {
+    fn write(&self, opt: WriteOpt) -> Option<String> {
+        let width = self.path.iter().map(|p| p.len() + 1).sum::<usize>() + self.name.len();
+        opt.consume_width(width as u16)?;
+
+        let mut r = String::new();
+        for part in &self.path {
+            r += &write_ident_part(part);
+            r += ".";
+        }
+        r += &write_ident_part(&self.name);
+        Some(r)
+    }
+}
+
+pub fn write_ident_part(s: &str) -> String {
+    fn forbidden_start(c: char) -> bool {
+        !(c.is_ascii_lowercase() || matches!(c, '_' | '$'))
+    }
+    fn forbidden_subsequent(c: char) -> bool {
+        !(c.is_ascii_lowercase() || c.is_ascii_digit() || matches!(c, '_'))
+    }
+    let needs_escape = s.is_empty()
+        || s.starts_with(forbidden_start)
+        || (s.len() > 1 && s.chars().skip(1).any(forbidden_subsequent));
+
+    if needs_escape {
+        format!("`{s}`")
+    } else {
+        s.to_string()
+    }
+}
```

Comment:
GPT4 wrote this as a regex:

```suggestion
use regex::Regex;
use once_cell::sync::Lazy;

static IDENT_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"^[_$a-z][_a-z\d]*$").unwrap());

pub fn write_ident_part(s: &str) -> String {
    if IDENT_REGEX.is_match(s) {
        s.to_string()
    } else {
        format!("`{}`", s)
    }
}
```

I actually though I did this before, but probably it was for something different...

Reply:
Should we also allow `*`, ref https://github.com/PRQL/prql/pull/2669/files#diff-107555a4ee015fdfa938922785f2d69c79432e2debb074e8270db0b05c9195c7R7 ?

Reply:
I remembered where I did this before — is this the same as https://github.com/prql/prql/blob/aa3ed6196596e38b3dfffad63d0b590e9eeed535/prql-compiler/src/sql/gen_expr.rs#L783 ? (fine to have it twice unless it's definitely going to remain the same)

Reply:
This is the problem with GPT: it looks likes it's correct, but it's not. The regex does not quote empty strings, but it should.

In anycase, this snippet was copied from the `gen_expr.rs` a while back and was living on `codegen` branch for last few months.

If you feel like it needs to be improved, go ahead. But add tests if you are using GPT.

Reply:
Very good point!! 

Yeah, agree it has that frailty. I have found it helpful for these sorts of "chore" tasks; agree tests become more important...

Reply:
Though in this case is it actually wrong? An empty string does get quoted by the regex.

Reply:
No, it's actually correct. An empty string does get quoted by the regex, which is correct. The logic has also been inverted.

But that's my point: it's more work validating the output as opposed to rewriting it. And things like this don't even need a rewrite or any more time spent on them.